libckteec: Key wrap unwrap

ruchi393 opened the pull request.

etienne-lms commented on libckteec/src/pkcs11_api.c:
```diff
@@ -1318,19 +1328,29 @@ CK_RV C_UnwrapKey(CK_SESSION_HANDLE hSession,
+		     CKR_TOKEN_WRITE_PROTECTED, CKR_USER_NOT_LOGGED_IN,
+		     CKR_UNWRAPPING_KEY_HANDLE_INVALID, CKR_WRAPPED_KEY_INVALID,
+		     CKR_UNWRAPPING_KEY_SIZE_RANGE, CKR_WRAPPED_KEY_LEN_RANGE,
+		     CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT);
```
For the 4 last line, prefer keep the alphabetical ordering, as in the spec:
```patch
		     CKR_TEMPLATE_INCOMPLETE, CKR_TEMPLATE_INCONSISTENT,
-		     CKR_TOKEN_WRITE_PROTECTED, CKR_USER_NOT_LOGGED_IN,
-		     CKR_UNWRAPPING_KEY_HANDLE_INVALID, CKR_WRAPPED_KEY_INVALID,
-		     CKR_UNWRAPPING_KEY_SIZE_RANGE, CKR_WRAPPED_KEY_LEN_RANGE,
-		     CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT);
+		     CKR_TOKEN_WRITE_PROTECTED,
+		     CKR_UNWRAPPING_KEY_HANDLE_INVALID,
+		     CKR_UNWRAPPING_KEY_SIZE_RANGE,
+		     CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT,
+		     CKR_USER_NOT_LOGGED_IN, CKR_WRAPPED_KEY_INVALID,
+		     CKR_WRAPPED_KEY_LEN_RANGE);

ruchi393 replied:
Changed. I was trying to optimize space usage :-). But you are right better to keep the same order. Makes comparison with spec easier.

etienne-lms commented on libckteec/src/pkcs11_processing.c:
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
+	buf += smecha.size;
+
+	/* Shm io2: output buffer reference - wrapped key */
+	if (key_len && *key_len) {
```
no need to test `key_len`

ruchi393 replied:
Removed

etienne-lms commented on libckteec/src/pkcs11_processing.c:
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
+	rv = ckteec_invoke_ctrl_out(PKCS11_CMD_WRAP_KEY, ctrl, out_shm,
+				    &out_size);
+
+	if (key_len && (rv == CKR_OK || rv == CKR_BUFFER_TOO_SMALL))
```
no need to test `key_len`

ruchi393 replied:
You are right. We have already checked it at beginning of the function. Have removed the check.

etienne-lms commented on libckteec/src/pkcs11_processing.c:
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
+	size_t out_size = 0;
+
+	if (!handle || !mechanism || (count && !attribs) ||
+	    (!unwrapping_key) || (wrapped_key_len && !wrapped_key))
```
can remove parentheses around `unwrapping_key`

ruchi393 replied:
Done. Travis also complained about it .

etienne-lms commented on libckteec/src/pkcs11_processing.c:
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
+bail:
+	ckteec_free_shm(out_shm);
+	ckteec_free_shm(ctrl);
+
```
release_serial_object(&smecha);

ruchi393 replied:
My bad. I have added it now.

etienne-lms reviewed (commented):
Review comments for "libckteec: Add implementation for wrapping/unwrapping keys".

`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
* "libckteec: Add missing indirect template"
* "libckteec: Add missing error codes"
* "libckteec: Allocate command ID's for wrapping/unwrapping keys"
  (s/ID's/IDs/ in commit log)
* "libckteec: Correct error code in C_Sign()"
  Maybe add tag  `Fixes: 3f44b8702995 ("libckteec: add support for Signing and Verification Functions") `

ruchi393 commented:
Thanks @etienne-lms . I have pushed changes to take care of review comments and error reported by Travis.

vesajaaskelainen commented on libckteec/src/pkcs11_processing.c:
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
+	TEEC_SharedMemory *ctrl = NULL;
+	TEEC_SharedMemory *out_shm = NULL;
+	struct serializer smecha = { 0 };
+	uint32_t session_handle = session;
```
CK_SESSION_HANDLE is CK_ULONG and that is 64 bits in 64 bit arch and does not fit in 32 uint32_t without a cast.

vesajaaskelainen replied:
We do have `serialize_ck_ulong()` but I suppose that is a bit off but we could have helper for `CK_ULONG`.

vesajaaskelainen replied:
I suppose it is OK to go with the existing way before the master is fixed for all of those?

vesajaaskelainen commented on libckteec/src/pkcs11_processing.c:
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
+	TEEC_SharedMemory *out_shm = NULL;
+	struct serializer smecha = { 0 };
+	uint32_t session_handle = session;
+	uint32_t wrp_key_handle = wrapping_key;
```
`CK_ULONG` issue here too

vesajaaskelainen commented on libckteec/src/pkcs11_processing.c:
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
+	struct serializer smecha = { 0 };
+	uint32_t session_handle = session;
+	uint32_t wrp_key_handle = wrapping_key;
+	uint32_t key_handle = key;
```
And here too.

vesajaaskelainen commented on libckteec/src/pkcs11_processing.c:
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
+	buf += sizeof(key_handle);
+
+	memcpy(buf, smecha.buffer, smecha.size);
+	buf += smecha.size;
```
I understand where you come from ;) -- but this is not needed.

ruchi393 replied:
Oh yes. There is nothing more to add in the buffer, so shifting this pointer is not required. I will remove this.

vesajaaskelainen commented:
Btw -- do we have CI build for OS side with 64 bit compiler? Because in theory it should have noticed that 64 bits do not fit in 32 bits.

vesajaaskelainen commented:
Btw. please tune up the PR's title -- we do not have TA's in optee client.

ruchi393 commented:
> Btw. please tune up the PR's title -- we do not have TA's in optee client.

Changed. Thanks for noticing this.

vesajaaskelainen commented on libckteec/include/pkcs11_ta.h:
```diff
@@ -657,6 +657,40 @@ enum pkcs11_ta_cmd {
+	 *
+	 * This command relates to the PKCS#11 API function C_WrapKey().
+	 */
+	PKCS11_CMD_WRAP_KEY = 50,
```
+1

vesajaaskelainen commented on libckteec/include/pkcs11_ta.h:
```diff
@@ -657,6 +657,40 @@ enum pkcs11_ta_cmd {
+	 *
+	 * This command relates to the PKCS#11 API function C_UnwrapKey().
+	 */
+	PKCS11_CMD_UNWRAP_KEY = 51,
```
+1

ruchi393 commented:
Rebased, squashed commits, tested again and updated the PR.

@etienne-lms , @vesajaaskelainen , please help review.

vesajaaskelainen commented on libckteec/src/pkcs11_processing.c:
```diff
@@ -1520,3 +1520,178 @@ CK_RV ck_generate_key_pair(CK_SESSION_HANDLE session,
+	memcpy(buf, &session_handle, sizeof(session_handle));
+	buf += sizeof(session_handle);
+
+	memcpy(buf, &obj_handle, sizeof(obj_handle));
```
`obj_handle` as a name do not have any relation that it is unwrapping key.

vesajaaskelainen commented on libckteec/src/pkcs11_processing.h:
```diff
@@ -127,4 +127,13 @@ CK_RV ck_generate_key_pair(CK_SESSION_HANDLE session,
+CK_RV ck_unwrap_key(CK_SESSION_HANDLE session, CK_MECHANISM_PTR mechanism,
+		    CK_OBJECT_HANDLE unnwrapping_key, CK_BYTE_PTR wrapped_key,
+		    CK_ULONG key_len, CK_ATTRIBUTE_PTR attribs, CK_ULONG count,
+		    CK_OBJECT_HANDLE_PTR handle);
```
In implementation you have nicer variable names -- please use them here too:
```
CK_RV ck_unwrap_key(CK_SESSION_HANDLE session, CK_MECHANISM_PTR mechanism,
		    CK_OBJECT_HANDLE unwrapping_key, CK_BYTE_PTR wrapped_key,
		    CK_ULONG wrapped_key_len, CK_ATTRIBUTE_PTR attribs,
		    CK_ULONG count, CK_OBJECT_HANDLE_PTR handle)
```

vesajaaskelainen commented on libckteec/src/pkcs11_processing.c:
```diff
@@ -1520,3 +1520,178 @@ CK_RV ck_generate_key_pair(CK_SESSION_HANDLE session,
+
+CK_RV ck_wrap_key(CK_SESSION_HANDLE session, CK_MECHANISM_PTR mechanism,
+		  CK_OBJECT_HANDLE wrapping_key, CK_OBJECT_HANDLE key,
+		  CK_BYTE_PTR wrapped_key, CK_ULONG_PTR key_len)
```
/key_len/wrapped_key_len/

vesajaaskelainen commented on libckteec/src/pkcs11_processing.h:
```diff
@@ -127,4 +127,13 @@ CK_RV ck_generate_key_pair(CK_SESSION_HANDLE session,
 
+CK_RV ck_wrap_key(CK_SESSION_HANDLE session, CK_MECHANISM_PTR mechanism,
+		  CK_OBJECT_HANDLE wrapping_key, CK_OBJECT_HANDLE key,
+		  CK_BYTE_PTR wrapped_key, CK_ULONG_PTR key_len);
```
/key_len/wrapped_key_len/

vesajaaskelainen commented:
`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`
- https://github.com/OP-TEE/optee_client/pull/265/commits/77408ed6798651b8f7d0c9d568575a81f188ebc5 - "libckteec: Add missing indirect template"
- https://github.com/OP-TEE/optee_client/pull/265/commits/f25a836f1609cdce9da5698067914a5893e39e9b - "libckteec: Add missing error codes"
- https://github.com/OP-TEE/optee_client/pull/265/commits/70a6b85175ec4b02340a4a3a5de8c7f5413acdfb - "libckteec: Allocate command ID's for wrapping/unwrapping keys"
- https://github.com/OP-TEE/optee_client/pull/265/commits/ec152f565049d25f19629322266619b1049f99aa - "libckteec: Correct error code in C_Sign()"

And with those minor things fixed:
- https://github.com/OP-TEE/optee_client/pull/265/commits/c95b48e5f0dc9004cca46b5cd95eaa0011633806 - "libckteec: Add implementation for wrapping/unwrapping keys"

ruchi393 commented:
Added tags as suggested.

Fixed "libckteec: Add implementation for wrapping/unwrapping keys" based on comments above.

Thanks @vesajaaskelainen for the reviewing.

vesajaaskelainen commented:
For commit:
- https://github.com/OP-TEE/optee_client/pull/265/commits/517ed320728114e68eb1d4aee2a71d536a8367eb - libckteec: Add implementation for wrapping/unwrapping keys

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

ruchi393 commented:
Rebased and tags applied.

jforissier commented:
I think this is safe to merge even though the `optee_os` part is still being reviewed, isn't it?

ruchi393 commented:
> I think this is safe to merge even though the `optee_os` part is still being reviewed, isn't it?

I would agree,we can merge this now.
